File: frontend/src/types/mpm.ts
export type TaskInput = { id: string; duration: number; predecessors: string[] }
export type TaskResult = TaskInput & { successors: string[]; earliest_start: number; earliest_finish: number; latest_start: number; latest_finish: number; total_float: number; free_float: number; is_critical: boolean }
export type GraphNode = { id: string; kind: 'start' | 'task' | 'finish'; earliest: number; latest: number; margin: number; duration: number; is_critical: boolean; level: number }
export type GraphEdge = { source: string; target: string; weight: number; is_critical: boolean }
export type MPMResult = { project_duration: number; tasks: TaskResult[]; nodes: GraphNode[]; edges: GraphEdge[]; critical_tasks: string[]; critical_paths: string[][] }

File: backend/app/schemas/mpm.py
from pydantic import BaseModel, Field, field_validator


class TaskInput(BaseModel):
    id: str = Field(min_length=1, max_length=80)
    duration: float = Field(ge=0)
    predecessors: list[str] = Field(default_factory=list)

    @field_validator("id")
    @classmethod
    def clean_id(cls, value: str) -> str:
        return value.strip()


class MPMRequest(BaseModel):
    tasks: list[TaskInput] = Field(min_length=1)


class TaskResult(BaseModel):
    id: str
    duration: float
    predecessors: list[str]
    successors: list[str]
    earliest_start: float
    earliest_finish: float
    latest_start: float
    latest_finish: float
    total_float: float
    free_float: float
    is_critical: bool


class GraphNode(BaseModel):
    id: str
    kind: str
    earliest: float
    latest: float
    margin: float
    duration: float
    is_critical: bool
    level: int


class GraphEdge(BaseModel):
    source: str
    target: str
    weight: float
    is_critical: bool


class MPMResponse(BaseModel):
    project_duration: float
    tasks: list[TaskResult]
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    critical_tasks: list[str]
    critical_paths: list[list[str]]

